lib/engine/tile.rb

```diff
@@ -190,11 +190,18 @@ def lawson?
         ((cities.empty? && towns.one?) && edges.size > 2)
     end
 
-    def upgrade_cost(abilities)
-      ability = abilities.find { |a| a.type == :tile_discount }
+    def upgrade_cost(entity, game)
```

Comment:
so this is a smell, upgrade_cost should be moved out of tile.rb and into the round next to border_cost